Change URI construction for Chant List and Chant Inventory

lucasmarchd01 opened the pull request:
This pull request modifies our url paths for the `chant-list` page from `/chants/?source=<source id` to `/source/<source id>/chants`. This makes it clear that the list of chants belongs to a specific source. A similar change is made for the `chant-index` page, which is now renamed to `chant-inventory`. Here is a summary of the changes:

- Change chant-list URI construction
- Change chant-index URI construction
- Modify templates to reflect changes in URI construction
- Update views and tests to reflect changes in URI construction

We should consider still setting up redirects for backwards compatibility.

Fixes #1202, Fixes #1201, Fixes #1210

ahankinson commented:
How many people's bookmarks will break with this change? Should the redirects be included in the change?

lucasmarchd01 commented:
> How many people's bookmarks will break with this change? Should the redirects be included in the change?

I'm not sure how many people would bookmark the chant-list or chant-inventory pages, but I suspect there might be some. It might make more sense to fix it in this PR rather than making a new issue. (Changing to draft)

jacobdgm commented:
> How many people's bookmarks will break with this change? Should the redirects be included in the change?

+1 here - redirects should definitely be included as part of this PR.

Especially if this redirect is a function-based view like our other redirect views, it should be really easy to return a `400 - Bad Request` if the user doesn't include a `?source=` parameter (allowing us to address #1210 at the same time).

ahankinson commented on django/cantusdb_project/main_app/urls.py:
```diff
@@ -72,7 +72,11 @@
     DifferentiaAutocomplete,
     ProvenanceAutocomplete,
     ProofreadByAutocomplete,
+    handle400,
```
I don't think you need these imports since you're not using these directly. You could change your code below to not express these as strings, but the way it is now they're unused.

ahankinson commented on django/cantusdb_project/main_app/views/views.py:
```diff
@@ -805,8 +806,14 @@ def redirect_node_url(request, pk: int) -> HttpResponse:
 
 
+def handle400(request, exception):
+    print(f"Status Code: {exception.status_code}")
```
Errant print statement. This should be removed.

ahankinson commented on django/cantusdb_project/main_app/views/views.py:
```diff
@@ -805,8 +806,14 @@ def redirect_node_url(request, pk: int) -> HttpResponse:
+
 def handle404(request, exception):
-    return render(request, "404.html")
+    print(f"Status Code: {exception.status_code}")
```
Also remove here.

ahankinson commented on django/cantusdb_project/main_app/views/views.py:
```diff
@@ -959,6 +966,36 @@ def redirect_documents(request) -> HttpResponse:
 
 
+def redirect_chant_list(request) -> HttpResponse:
+    source_id: str = request.GET.get("source")
```
```suggestion
    source_id: Optional[str] = request.GET.get("source")
```

ahankinson commented on django/cantusdb_project/main_app/views/views.py:
```diff
@@ -959,6 +966,36 @@ def redirect_documents(request) -> HttpResponse:
+    }
+
+    query_string: str = urlencode(params)
+    url: str = "{}{}".format(base_url, f"?{query_string}" if query_string else "")
```
It's a bit weird to mix `format()` calls with f-string calls.

ahankinson commented on django/cantusdb_project/main_app/views/views.py:
```diff
@@ -959,6 +966,36 @@ def redirect_documents(request) -> HttpResponse:
+    folio: str = request.GET.get("folio")
+    search_text: str = request.GET.get("search_text")
+
+    params: dict = {
```
You should move the dictionary construction outside of the comprehension. It's clearer, and actually takes fewer lines.

```
d = {
  "feast": feast_id,
  "genre": genre_id,
  "folio": folio,
  "search_text": search_text,
}
params = {k: v for k, v in d.items() if v is not None}

ahankinson commented on django/cantusdb_project/main_app/views/views.py:
```diff
@@ -959,6 +966,36 @@ def redirect_documents(request) -> HttpResponse:
+    base_url: str = reverse("chant-list", args=[source_id])
+
+    # optional search params
+    feast_id: str = request.GET.get("feast")
```
These should all be `Optional[str]`

```suggestion
    feast_id: Optional[str] = request.GET.get("feast")
```

A good rule is that if you need to check if it's `None` later, you must have marked it as Optional.

(Are you using any sort of type-checker, like mypy, when you develop? It helps catch these things. Also a good IDE will flag any unused imports, etc.)

lucasmarchd01 commented:
Thanks @ahankinson and @jacobdgm for the helpful suggestions!

In this update, I removed the custom Django error handlers. After thorough testing, it was observed that these handlers were not being called, and Django is able to find and display error pages manually.

Additionally, our application doesn't have any specific requirements or special handling that necessitates custom error handlers. The removal of these custom handlers has no impact on the application's error handling capabilities, and Django is able to handle errors effectively without them.

I have also added tests for the chant-list and chant-inventory redirects.

ahankinson commented on django/cantusdb_project/main_app/urls.py:
```diff
@@ -73,6 +73,8 @@
     ProvenanceAutocomplete,
     ProofreadByAutocomplete,
 )
+from django.conf.urls import handler400, handler404
```
Are these used?

ahankinson replied:
I see you’ve removed them — great! If you’re not using a code editor that highlights these sort of syntax or style errors, you should. They are widely available and help catch these sorts of things before it needs “human linting”

ahankinson commented on django/cantusdb_project/main_app/templates/400.html:
```diff
@@ -0,0 +1,39 @@
+{% load static %}
+<html lang="en">
```
Is this file still used too?

lucasmarchd01 replied:
Our static file is used to load the background image displayed on the custom error pages.

dchiller replied:
I am also a bit confused by this. It looks like `400.html` is introduced with this PR, but not referenced as far as I can tell anywhere else in the changes? Does django look for a `400.html` by default? 

Updated: It seems like django doesn't and that a view would need to be created and set as `handler400` in `urls.py`. https://docs.djangoproject.com/en/5.0/ref/urls/#module-django.conf.urls

Updated again: It does...thanks @lucasmarchd01

lucasmarchd01 replied:
I spent a while verifying what is happening with this before removing the error handlers, and found that when Django encounters a 400 error (or 404, etc.) it will locate the 400.html page in the template directory by default. The error handlers are for the case where you might want to customize the error view. Django is able to serve the correct html files even if they are not referenced.

https://docs.djangoproject.com/en/5.0/ref/views/#the-404-page-not-found-view

dchiller commented on django/cantusdb_project/main_app/views/views.py:
```diff
@@ -959,6 +956,43 @@ def redirect_documents(request) -> HttpResponse:
+    query_string: str = urlencode(params)
+    url: str = f"{base_url}?{query_string}" if query_string else base_url
+
+    return redirect(url)
```
this issues a temporary redirect by default, but I think this is intended to be permanent?

ahankinson replied:
Good catch

lucasmarchd01 replied:
Yes thanks!

jacobdgm commented on django/cantusdb_project/main_app/tests/test_views.py:
```diff
@@ -2917,6 +2934,22 @@ def test_dd_column(self):
+        self.assertRedirects(response, reverse("chant-inventory", args=[source_id]))
+
+    def test_redirect_without_source_parameter(self):
+        url = reverse("redirect-chant-index")
```
The main place this needs to be changed is in `urls.py`, but this view should be `"redirect-chant-inventory", shouldn't it?

More generally, shouldn't this view now be called `SourceInventoryView`, and live in `views/source.py`? (Since/if this is the case, we maybe still want the redirect view to be "redirect-chant-index" or "-inventory", since we're redirecting from the old way the view was set up, with `/chant/` in the URL path...)

jacobdgm commented on django/cantusdb_project/main_app/urls.py:
```diff
@@ -293,6 +293,11 @@
     ),
+    path(
+        "index/",
+        views.redirect_index_list,
```
Not a fan of `redirect_index_list` as the name of the view....

jacobdgm reviewed (approved):
I think it's within the scope of this PR to rename the affected views to something that more accurately indicates what they do.

dchiller commented on django/cantusdb_project/main_app/urls.py:
```diff
@@ -491,5 +501,3 @@
         name="differentia-autocomplete",
     ),
 ]
-
```
Also, with the removal of this, will the custom 404.html be used?

jacobdgm commented on django/cantusdb_project/main_app/tests/test_views.py:
```diff
@@ -5539,7 +5573,7 @@ def test_incipit_search(self):
         chant_with_normal_cantus_id = make_fake_chant(cantus_id="012345")
         chant_with_numerals_in_incipit = make_fake_chant(
-            incipit="0 me! 0 my! This is unexpected!"
+            cantus_id="223455", incipit="0 me! 0 my! This is unexpected!"
```
Is this Black compliant?

Also, this diff seems to be comparing to a not fully up-to-date `develop` branch... I don't know why this should be the case. Maybe GitHub will recalculate things when this PR stops being a draft?

dchiller replied:
speaking of which...note that black's 2024 stable style guide was released in the last few days, so an upgrade might be in order.

lucasmarchd01 replied:
> Is this Black compliant?
> 
> Also, this diff seems to be comparing to a not fully up-to-date `develop` branch... I don't know why this should be the case. Maybe GitHub will recalculate things when this PR stops being a draft?

The potential conflict should be resolved now.

ahankinson commented on django/cantusdb_project/main_app/templates/chant_detail.html:
```diff
@@ -267,10 +267,10 @@ <h4>List of melodies</h4>
                                 {% endif %}
                                 {% if next_folio %}
-                                    &nbsp;<a href="{% url "chant-list" %}?source={{ source.id }}&folio={{ next_folio }}">> {{ next_folio }}</a>
+                                    &nbsp;<a href="{% url "chant-list" source.id %}?folio={{ next_folio }}">> {{ next_folio }}</a>
```
You can use keyword arguments in the `url` template tag so that you don't need to do string concatenation here...
